feat: add C implementation for `stats/base/dists/rayleigh/skewness`

DhruvArvindSingh opened the pull request:
Resolves #3809 .

## Description

> feat: add C implementation for `@stdlib/stats/base/dists/rayleigh/skewness`

This pull request:

-   #3809 

## Related Issues

> #3809 

This pull request:

-   resolves #3809  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> ![Screenshot 2024-12-27 182822](https://github.com/user-attachments/assets/21d048e8-8029-41fc-9ada-7ef05cbe062a)
> ![Screenshot 2024-12-27 182835](https://github.com/user-attachments/assets/9c34b42c-09d8-4a94-8bd8-835006d7f00d)



No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: add C implementation for @stdlib/stats/base/dists/rayleigh/skewness'

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh commented:
Hey @kgryte do review this

DhruvArvindSingh committed:
fix(readme): updated readme.md

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/stats/base/dists/rayleigh/skewness.h"
+#include "stdlib/math/base/napi/unary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/src/main.c:
```diff
@@ -0,0 +1,58 @@
+* @param {NonNegativeNumber} sigma - scale parameter
+* @returns {NonNegativeNumber} skewness
+*
+* @example
+* var v = skewness( 9.0 );
+* // returns ~0.631
+*
+* @example
+* var v = skewness( 2.0 );
+* // returns ~0.631
+*
+* @example
+* var v = skewness( -0.2 );
+* // returns NaN
+*
+* @example
+* var v = skewness( NaN );
+* // returns NaN
```
```suggestion
* @example
* double v = skewness( 9.0 );
* // returns ~0.631
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/src/main.c:
```diff
@@ -0,0 +1,58 @@
+/**
+* Returns the skewness of a Rayleigh distribution.
+*
+* @param {NonNegativeNumber} sigma - scale parameter
+* @returns {NonNegativeNumber} skewness
```
```suggestion
* @param sigma   scale parameter
* @return        skewness
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,139 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		sigma[ i ] = random_uniform( 0.0, 20.0 ) + STDLIB_CONSTANT_FLOAT64_EPS;
```
```suggestion
		sigma[ i ] = random_uniform( STDLIB_CONSTANT_FLOAT64_EPS, 20.0 );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/README.md:
```diff
@@ -116,6 +116,99 @@ for ( i = 0; i < 10; i++ ) {
+    for ( i = 0; i < 25; i++ ) {
+        sigma = random_uniform( 0.0, 10.0 );
+        y = stdlib_base_dists_rayleigh_skewness( sigma );
+        printf( "σ: %lf, f(x;σ): %lf\n", sigma, y );
```
```suggestion
        printf( "σ: %lf, skew(X,σ): %lf\n", sigma, y );
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/src/main.c:
```diff
@@ -0,0 +1,46 @@
+	) {
+		return 0.0/0.0; // NaN
+	}
+	SKEWNESS = 2.0 * STDLIB_CONSTANT_FLOAT64_SQRT_PI * ( STDLIB_CONSTANT_FLOAT64_PI-3.0 ) / stdlib_base_pow( 4.0 - STDLIB_CONSTANT_FLOAT64_PI, 1.5 );
```
We should fully precompute this constant and declare it in L24.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/src/main.c:
```diff
@@ -0,0 +1,46 @@
+* @return        skewness
+*
+* @example
+* double v = skewness( 9.0 );
```
```suggestion
* double v = stdlib_base_dists_rayleigh_skewness( 9.0 );
```

Planeshifter committed:
chore: apply suggestions from code review

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/benchmark/benchmark.js:
```diff
@@ -32,13 +33,19 @@ var skewness = require( './../lib' );
+	len = 100;
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		sigma[ i ] = ( randu()*20.0 ) + EPS;
```
Should use `uniform` to draw directly from desired distribution.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,68 @@
+	len = 100;
+	sigma = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		sigma[ i ] = ( randu()*20.0 ) + EPS;
```
Should use `uniform` to draw directly from desired distribution.

Planeshifter committed:
chore: precompute constant and use uniform

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: add newline and use uniform

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
fix: use correct value for constant

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/README.md:
```diff
@@ -116,6 +116,99 @@ for ( i = 0; i < 10; i++ ) {
+
+#### stdlib_base_dists_rayleigh_skewness( sigma )
+
+Evaluates the skewness for a Rayleigh distribution.
```
```suggestion
Returns the [skewness][skewness] of a [Rayleigh][rayleigh-distribution] distribution with scale parameter `sigma`.
```

Planeshifter committed:
docs: update description

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/manifest.json:
```diff
@@ -0,0 +1,83 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/binary",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/constants/float64/pi"
```
```suggestion
        "@stdlib/math/base/napi/binary",
        "@stdlib/math/base/assert/is-nan"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/manifest.json:
```diff
@@ -0,0 +1,83 @@
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/constants/float64/eps",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/constants/float64/pi"
```
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/constants/float64/eps"
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/manifest.json:
```diff
@@ -0,0 +1,83 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/constants/float64/pi"
```
```suggestion
        "@stdlib/math/base/assert/is-nan"
```

Planeshifter committed:
fix: remove extra dependencies from manifest.json

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (dismissed):
Looks good to me now after making the required edits.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/rayleigh/skewness/manifest.json:
```diff
@@ -0,0 +1,77 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/binary",
```
```suggestion
        "@stdlib/math/base/napi/unary",
```

Planeshifter committed:
fix: use correct napi dependency

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
